Cannot push port (Balance.out8) twice

I'm getting this error using Akka Streams 2.4.8. I am not sure where it comes from, the stacktrace does not tell much:

```
[ERROR] [07/12/2016 14:02:27.093] [default-akka.actor.default-dispatcher-248] [akka://default/user/StreamSupervisor-18/flow-18-21-unknown-operation] Error in stage [Balance]: requirement failed: Cannot push port (Balance.out8) twice
java.lang.IllegalArgumentException: requirement failed: Cannot push port (Balance.out8) twice
        at scala.Predef$.require(Predef.scala:219)
        at akka.stream.stage.GraphStageLogic.push(GraphStage.scala:436)
        at akka.stream.scaladsl.Balance$$anon$17.akka$stream$scaladsl$Balance$$anon$$dequeueAndDispatch(Graph.scala:613)
        at akka.stream.scaladsl.Balance$$anon$17$$anon$18.onPush(Graph.scala:618)
        at akka.stream.impl.fusing.GraphInterpreter.processElement$1(GraphInterpreter.scala:590)
        at akka.stream.impl.fusing.GraphInterpreter.processEvent(GraphInterpreter.scala:601)
        at akka.stream.impl.fusing.GraphInterpreter.execute(GraphInterpreter.scala:542)
        at akka.stream.impl.fusing.GraphInterpreterShell.runBatch(ActorGraphInterpreter.scala:471)
        at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:410)
        at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:603)
        at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:618)
        at akka.actor.Actor$class.aroundReceive(Actor.scala:484)
        at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:529)
        at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
        at akka.actor.ActorCell.invoke(ActorCell.scala:495)
        at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
        at akka.dispatch.Mailbox.run(Mailbox.scala:224)
        at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
        at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
        at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
        at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
        at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

I am using a [balancer](http://doc.akka.io/docs/akka/2.4.8/scala/stream/stream-cookbook.html#cookbook-balance-scala) to distribute lines processing from a file source.

--- a/akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBalanceSpec.scala
+++ b/akka-stream-tests/src/test/scala/akka/stream/scaladsl/GraphBalanceSpec.scala
@@ -3,10 +3,11 @@ package akka.stream.scaladsl
 import scala.concurrent.Await
 import scala.concurrent.duration._
 import scala.concurrent.Future
-import akka.stream.{ SourceShape, ClosedShape, ActorMaterializer, ActorMaterializerSettings }
+import akka.stream._
 import akka.stream.testkit._
 import akka.stream.testkit.scaladsl._
 import akka.stream.testkit.Utils._
+import akka.util.ByteString
 
 class GraphBalanceSpec extends StreamSpec {
 
@@ -256,6 +257,36 @@ class GraphBalanceSpec extends StreamSpec {
       bsub.expectCancellation()
     }
 
+    // Bug #20943
+    "not push output twice" in assertAllStagesStopped {
+      val p1 = TestPublisher.manualProbe[Int]()
+      val c1 = TestSubscriber.manualProbe[Int]()
+      val c2 = TestSubscriber.manualProbe[Int]()
+
+      RunnableGraph.fromGraph(GraphDSL.create() { implicit b ⇒
+        val balance = b.add(Balance[Int](2))
+        Source.fromPublisher(p1.getPublisher) ~> balance.in
+        balance.out(0) ~> Sink.fromSubscriber(c1)
+        balance.out(1) ~> Sink.fromSubscriber(c2)
+        ClosedShape
+      }).run()
+
+      val bsub = p1.expectSubscription()
+      val sub1 = c1.expectSubscription()
+      val sub2 = c2.expectSubscription()
+
+      sub1.request(1)
+      p1.expectRequest(bsub, 16)
+      bsub.sendNext(1)
+      c1.expectNext(1)
+
+      sub2.request(1)
+      sub2.cancel()
+      bsub.sendNext(2)
+
+      sub1.cancel()
+      bsub.expectCancellation()
+    }
   }
 
 }

--- a/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala
+++ b/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala
@@ -610,8 +610,16 @@ final class Balance[T](val outputPorts: Int, waitForAllDownstreams: Boolean) ext
 
     private def dequeueAndDispatch(): Unit = {
       val out = pendingQueue.dequeue()
-      push(out, grab(in))
-      if (!noPending) pull(in)
+      if (out ne null) {
+        // the out triggering the pull may since closed
+        if (!isClosed(out)) {
+          push(out, grab(in))
+          if (!noPending) pull(in)
+        } else {
+          // try to find one that isn't closed in that case
+          dequeueAndDispatch()
+        }
+      }
     }
 
     setHandler(in, new InHandler {

--- a/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala
+++ b/akka-stream/src/main/scala/akka/stream/scaladsl/Graph.scala
@@ -608,15 +608,17 @@ final class Balance[T](val outputPorts: Int, waitForAllDownstreams: Boolean) ext
     private var needDownstreamPulls: Int = if (waitForAllDownstreams) outputPorts else 0
     private var downstreamsRunning: Int = outputPorts
 
+    @tailrec
     private def dequeueAndDispatch(): Unit = {
       val out = pendingQueue.dequeue()
+      // out is null if depleted pendingQueue without reaching
+      // an out that is not closed, in which case we just return
       if (out ne null) {
-        // the out triggering the pull may since closed
         if (!isClosed(out)) {
           push(out, grab(in))
           if (!noPending) pull(in)
         } else {
-          // try to find one that isn't closed in that case
+          // try to find one output that isn't closed
           dequeueAndDispatch()
         }
       }

--- a/akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala
+++ b/akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala
@@ -453,8 +453,8 @@ abstract class GraphStageLogic private[stream] (val inCount: Int, val outCount: 
 
       // Detailed error information should not add overhead to the hot path
       ReactiveStreamsCompliance.requireNonNullElement(elem)
-      require(isAvailable(out), s"Cannot push port ($out) twice")
       require(!isClosed(out), s"Cannot pull closed port ($out)")
+      require(isAvailable(out), s"Cannot push port ($out) twice")
 
       // No error, just InClosed caused the actual pull to be ignored, but the status flag still needs to be flipped
       connection.portState = portState ^ PushStartFlip

